src/topology/flush_control.rs

```diff
@@ -0,0 +1,274 @@
+use crate::bounded_channel::{BoundedReceiver, BoundedSender, bounded};
+use std::sync::{Arc, Mutex};
+use tokio::sync::broadcast;
+use tokio::sync::broadcast::{Receiver, Sender};
+use tower::BoxError;
+use tracing::warn;
+
+const FLUSH_CHAN_SIZE: usize = 20;
+
+#[derive(Debug, Clone)]
+pub struct FlushRequest {
+    id: u64,
+}
+
+#[derive(Debug, Clone)]
+pub struct FlushResponse {
+    id: u64,
+}
+
+pub struct FlushBroadcast {
+    inner: Arc<Mutex<Inner>>,
+
+    req_tx: Sender<FlushRequest>,
+    resp_rx: BoundedReceiver<FlushResponse>,
+    resp_tx: BoundedSender<FlushResponse>,
+}
+
+struct Inner {
+    listeners: usize,
+
+    _req_rx: Receiver<FlushRequest>, // must keep open, but we subscribe from tx to create listeners
+}
+
+impl Inner {
+    fn new(req_rx: Receiver<FlushRequest>) -> Self {
+        Self {
+            listeners: 0,
+            _req_rx: req_rx,
+        }
+    }
+
+    pub fn add_subscriber(&mut self) {
+        self.listeners += 1;
+    }
+}
+
+impl FlushBroadcast {
+    pub fn new() -> Self {
+        let (req_tx, req_rx) = broadcast::channel(FLUSH_CHAN_SIZE);
+        let (resp_tx, resp_rx) = bounded(FLUSH_CHAN_SIZE);
+
+        Self {
+            inner: Arc::new(Mutex::new(Inner::new(req_rx))),
+            req_tx,
+            resp_tx,
+            resp_rx,
+        }
+    }
+
+    pub fn into_parts(self) -> (FlushSender, FlushSubscriber) {
+        let publisher = FlushSender {
+            inner: self.inner.clone(),
+            next_req_id: 1,
+            req_tx: self.req_tx.clone(),
+            resp_rx: self.resp_rx,
+        };
+        let subscriber = FlushSubscriber {
+            inner: self.inner.clone(),
+            req_tx: self.req_tx.clone(),
+            resp_tx: self.resp_tx,
+        };
+
+        (publisher, subscriber)
+    }
+}
+
+pub struct FlushSender {
+    inner: Arc<Mutex<Inner>>,
+    next_req_id: u64,
+    req_tx: Sender<FlushRequest>,
+    resp_rx: BoundedReceiver<FlushResponse>,
+}
+
+impl FlushSender {
+    pub async fn broadcast(&mut self) -> Result<(), BoxError> {
+        let curr_listeners = self.inner.lock().unwrap().listeners;
+        let req_id = self.next_req_id;
+        self.next_req_id += 1;
+        let req = FlushRequest { id: req_id };
+
+        if let Err(e) = self.req_tx.send(req) {
+            return Err(format!("Unable to send broadcast message: {}", e).into());
+        }
+
+        let mut acked = 0u64;
+        loop {
```

Comment:
If for some reason the ack send failed on line 152 would we loop forever with the lock? If so, it seems unlikely but perhaps in the future we should add some timeout here just in case?

Reply:
Good call, I can probably drop the lock after reading the current listener count.

Reply:
Actually, looking at this the lock should be dropped immediately after reading the listeners count AFAICT.

Reply:
Yeah agreed, you're not keeping a reference to the guard so it's dropped immediately so that is cool 👍. I think you could still loop indefinitely here though if you don't get your ack, so you might want to add a timeout like you did on the other side at some point to? Kind of an edge case but could be useful to know in the logs if it ever were to occur. Debugging lambda stuff is already a janky challenge.

Reply:
Oh yes, so I'm currently calling the entire broadcast() method with a timeout from the lambda extension for that reason: https://github.com/streamfold/rotel-lambda-extension/blob/7b52788fb7ec5eb60bb184b47adbc798cd088d08/src/main.rs#L313-L315. I guess the timeout could be pushed into broadcast iteself, but I may just leave a clear comment on this for the time being.